Source/SwiftLintFramework/Rules/Lint/AccessibilityLabelForImageRule.swift

```diff
@@ -0,0 +1,103 @@
+/// however, an `Image` is an accessibility element by default. If the developer does not explicitly hide them from
+/// accessibility or give them an accessibility label, they will inherit the name of the image file, which often creates
+/// a poor experience when VoiceOver reads things like "close icon white".
+public struct AccessibilityLabelForImageRule: ConfigurationProviderRule, AutomaticTestableRule {
+    public var configuration = SeverityConfiguration(.warning)
+
+    public init() {}
+
+    public static let description = RuleDescription(
+        identifier: "accessibility_label_for_image",
+        name: "Accessibility Label for Image",
+        description: "All Images should either be hidden from accessibility or provide an accessibility label.",
+        kind: .lint,
+        minSwiftVersion: .fiveDotOne,
+        nonTriggeringExamples: [
+            Example("""
+            Image(decorative: "my-image")
+            """),
+            Example("""
+            Image("my-image")
+                .accessibility(hidden: true)
+            """),
+            Example("""
+            Image("my-image")
+                .accessibilityHidden(true)
+            """),
+            Example("""
+            Image("my-image")
+                .accessibility(label: Text("Alt text for my image")
+            """),
+            Example("""
+            Image("my-image")
+                .accessibilityLabel(Text("Alt text for my image"))
+            """),
+            Example("""
+            Image(uiImage: myUiImage)
+                .renderingMode(.template)
+                .foregroundColor(.blue)
+                .accessibilityHidden(true)
+            """),
+            Example("""
+            Image(uiImage: myUiImage)
+                .accessibilityLabel(Text("Alt text for my image"))
+            """),
+            Example("""
+            SwiftUI.Image(uiImage: "my-image").resizable().accessibilityHidden(true)
+            """)
+        ],
+        triggeringExamples: [
+            Example("""
+            ↓Image("my-image")
+                .resizable(true)
+                .frame(width: 48, height: 48)
+            """),
+            Example("""
+            ↓Image(uiImage: myUiImage)
+            """),
+            Example("""
+            SwiftUI.↓Image(uiImage: "my-image").resizable().accessibilityHidden(false)
+            """),
+            Example("""
+            Image(uiImage: myUiImage)
+                .resizable()
+                .frame(width: 48, height: 48)
+                .accessibilityLabel(Text("Alt text for my image"))
+            ↓Image("other image")
+            """),
+            Example("""
+            Image(decorative: "image1")
+            ↓Image("image2")
+            Image(uiImage: "image3")
+                .accessibility(label: Text("a pretty picture"))
+            """)
+        ]
+    )
+
+    private var pattern: String {
```

Comment:
Can you avoid the use of a regular expression for this rule? Regular expressions are great for custom rules when you don't have the fully programmatic environment and full access to the AST, but with native lint rules you do have access to that. It will be more performant and less likely to have false positives.

Reply:
Sure can definitely take a look at that @jpsim ! Any existing rules that I should look at for examples of how I can use the AST to check things like an `Image` literal with/without certain modifiers? My current level of understanding with linters is more on the regex level and I'm not super comfortable or confident working with the AST.

While there are some false positives with this, the majority of the violations this caught in the repos that were scanned are true positives — places where VoiceOver would read the image file name. This is an increasingly common accessibility issue as more apps are adopting SwiftUI, so would love to get some check for it into swiftlint, even if just as a warning level opt-in regex rule for now!

Reply:
Totally agree on wanting to keep things performant. I'll take a look at those and see what I can do 👍

Reply:
I'm here to help if you get stuck.

Reply:
Hey @jpsim , I finally got around to revisiting this and taking a stab at writing it as an ASTRule, but managed to mess up the rebase pretty badly (I think I may have rebase origin/master from my fork instead of upstream/master, or potentially rebased origin/master from my branch rather than the other way around... 🤦‍♂️) so I went ahead and created a new branch, cherry-picked the commits, and opened up #3862 to replace this.